In [1]:
# -*- coding: utf-8 -*-

# Unhuman Project #

## Tutorial - Keras ##

In this document, I will write the python code given by the [TensorFlow Tutorial: Keras Guide](https://www.tensorflow.org/guide/keras)

Keras is an API to help build and train artificial intelligence model, implemented in TensorFlow library. The *model* is a group of layers which contains neurons.

### Get Started with Keras ###

In Python, to import keras, the following code must be added:

In [2]:
import tensorflow as tf
import keras

Using TensorFlow backend.


We can now use keras through ```tf.keras``` module.

Note that you might need to install the ```keras``` python package by using:
```bash
pip install keras
```
or
```bash
conda install -n myenv keras
```

In [3]:
print("TensorFlow version: {0}\nKeras version: {1}".format(tf.__version__, keras.__version__))

TensorFlow version: 1.8.0
Keras version: 2.2.0


### Build Simple Model ###

#### Sequential Model ####

The following code build a Sequential model, with a fully-connected neural network (multi-layer perceptron):

In [4]:
model = keras.Sequential()

# Add a dense layer containing 64 neurons
model.add(keras.layers.Dense(units=64, activation="relu", input_shape=(32,)))

# Add another
model.add(keras.layers.Dense(units=64, activation="relu"))

# Add a softmax layer with 10 output neurons
model.add(keras.layers.Dense(units=10, activation="softmax"))

As a reminder, The ```relu``` activation function, also called **Rectifer function** returns 0 if the inputs ```x``` is less or equal to zero, or ```x``` otherwise. In other words, the function will conserve the effect during the training process. The following formula shows the function in a more mathematical way:
\begin{equation*}
f(x) = \begin{Bmatrix}
0 & if x \leq 0\\ 
x & if x > 0
\end{Bmatrix} = max(0, x)
\end{equation*}
![Retifier function](../../../../../res/images/relu.png)

---

The **Softmax function**, on the other hand, help to train classification model. The function $\sigma(z)_{j}$ takes as input a vector. The parameter $j$ is the index of the output neuron ($j = 1, 2, ..., K$). The following equation represents the Softmax function:
\begin{equation*}
\sigma(z)_{j} = \frac{e^{z_{j}}}{\sum_{k=1}^{K} e^{z_k}}
\end{equation*}

---

Another known activation function which is not in this python code is **Sigmoid function**. This function is widely use in artificial intelligence, and it is one of the most activation functions.
\begin{equation*}
f(x) = \frac{1}{1 + e^{-x}}
\end{equation*}
![Sigmoid function](../../../../../res/images/sigmoid.png)

#### Configuration ####

Amongst all [```tf.keras.layers```](https://www.tensorflow.org/api_docs/python/tf/keras/layers) constructors, there are common arguments that must be detailed before delve into it:
* **```activation=```:** The activation function of the neurons for this specific layer. It can be a function name (```str```) or the built-in function it-self (ex: ```tf.keras.layers.Dense(units=10, activation=tf.sigmoid)```).
* **```kernel_initialize=```:** The layer's weights at initialization, for every neuron in the layer (ex: ```kernel_initializer="orthogonal"```).
* **```biad_initialize=```:** The layer's bias weight only at initialization, for every neuron in the layer (ex: ```bias_initializer=tf.keras/initializers.constant(2.0)```).
* **```kernel_regularizer=```:** The regularization that modify the layer's weights (ex: ```kernel_regularizer=tf.keras.regularizers.l1(0.01)```). By default, there is no regularization.
* **```bias_regularizer=```:** The regularization that modify the layer's bias weight (ex: ```bias_regularizer=tf.keras.regularizers.l2(0.01)```). By default, there is no regularization.

### Training ###

#### Configuration ####

The method [```tf.keras.Model.compile```](https://www.tensorflow.org/api_docs/python/tf/keras/Model#compile) allow the developer to configure the training process. The most relevant arguments are:
* **```optimzer=```:** The optimizer that will train the model. See the module [```tf.train```](https://www.tensorflow.org/api_docs/python/tf/train) for more detail, and have a look at [```AdamOptimizer```](https://www.tensorflow.org/api_docs/python/tf/train/AdamOptimizer), [```RMSPropOptimizer```](https://www.tensorflow.org/api_docs/python/tf/train/RMSPropOptimizer) and [```GradientDescentOptimizer```](https://www.tensorflow.org/api_docs/python/tf/train/GradientDescentOptimizer) which are the three most known optimizer in TensorFlow.
* **```loss=```:** The function that compute the error between the computed value by the model, and the expected value (label). The code below is the error function in the [TensorFlow eager tutorial](https://www.tensorflow.org/tutorials/eager/custom_training_walkthrough#define_the_loss_and_gradient_function) that can be passed as argument for this parameter:
```python
def error(model, x, y):
	"""
	Compute the error between the value returned by model(x) and the expected result y
	:param model: The model to use
	:param x: The input(s)
	:param y: The expected value that the model is supposed to return with x as input(s)
	:return: The error between the actual value and the expected value
	"""
	computed_y = model(x)
	return tf.losses.sparse_softmax_cross_entropy(labels=y, logits=computed_y)
```
* **```metrics=```:** The metric to monitor and evaluate during the training/testing process. For more detail about this argument, please [see the documentation about it](https://www.tensorflow.org/api_docs/python/tf/keras/Model#compile). This parameter accept string, list of string, dictionary (containing string as value) or function type (which are listed [here](https://www.tensorflow.org/api_docs/python/tf/keras/metrics)). A typical example is ```metrics=["accuracy"]``` to monitor the accracy level, or ```metrics=["mae"]``` to watch the *mean absolute error*.

For more examples about this method, please have a look at the [TensorFlow tutorial page](https://www.tensorflow.org/guide/keras#set_up_training).

Now, let's configure the model:

In [5]:
model.compile(
    optimizer=tf.train.AdamOptimizer(0.001),
    loss="categorical_crossentropy",
    metrics=["accuracy"]
)

#### *Fit* Training ####

For this example, a small dataset will be used to train the model (as it doesn't have any purpose). Fortunatly, TensorFlow has implemented a function for that instead of passing through a training-loop. And this alternative is [```model.fit()```](https://www.tensorflow.org/api_docs/python/tf/keras/Model#fit) methods, which take 5 important parameters:
* **```x=```:** The inputs (*features*) (an array of array of int, for instance)
* **```y=```:** The expected outputs (*labels*)
* **```batch_size=```:** The number of *slices* of inputs to cut for each iteration. A typical value is ```32```. Warning: The last batch might be very small if the number of samples is not divisible by ```batch_size```.
* **```epochs=```:** Number of iterations through the whole dataset input.
* **```verbose=```:** (*Optional*) Set the log level: 0 for quiet mode, 1 to display a progressbar (default), or 2 for one line per epoch.
* **```validation_data=```:** (*Optional*) Tuple of features and labels. The ```fit``` function will compute the error made by the model at each iteration using this dataset and print it.

Time for training the model. I will use the library [NumPy](https://www.numpy.org/) to generate a random dataset:

In [6]:
import numpy as np

features = np.random.random((1000, 32)) # Generate an array containing 1000 array which contain 32 values each
labels   = np.random.random((1000, 10)) # array containing 1000 arrays containing 10 values

val_features = np.random.random((1000, 32))
val_labels   = np.random.random((1000, 10))

model.fit(
    x=features,
    y=labels,
    batch_size=32,
    epochs=10,
    validation_data=(val_features, val_labels)
)

Train on 1000 samples, validate on 1000 samples
Epoch 1/10
1000/1000 [==============================] - 0s 347us/step - loss: 11.6075 - acc: 0.1150 - val_loss: 11.6078 - val_acc: 0.1060
Epoch 2/10
1000/1000 [==============================] - 0s 94us/step - loss: 11.5483 - acc: 0.1110 - val_loss: 11.6036 - val_acc: 0.1050
Epoch 3/10
1000/1000 [==============================] - 0s 98us/step - loss: 11.5406 - acc: 0.1310 - val_loss: 11.6025 - val_acc: 0.0960
Epoch 4/10
1000/1000 [==============================] - 0s 98us/step - loss: 11.5343 - acc: 0.1270 - val_loss: 11.6026 - val_acc: 0.1020
Epoch 5/10
1000/1000 [==============================] - 0s 103us/step - loss: 11.5302 - acc: 0.1360 - val_loss: 11.6021 - val_acc: 0.1080
Epoch 6/10
1000/1000 [==============================] - 0s 112us/step - loss: 11.5243 - acc: 0.1450 - val_loss: 11.6034 - val_acc: 0.0960
Epoch 7/10
1000/1000 [==============================] - 0s 108us/step - loss: 11.5203 - acc: 0.1550 - val_loss: 11.6023 - val_a

If you execute this code, you must see above the output given by the ```fit``` function after computing the validation data at each iteration. You may found a very high loss value (~11.5) and an accuracy value very low (~0.1). The answer of this paradox is that the training dataset and the validation dataset are both **randomly generated**, and then they might not have any link between those two sets. So don't worry if you find such values.

Now let's see how to train the model with larger dataset. In this case, the ```fit``` function must take the class [```tf.data.Dataset```](https://www.tensorflow.org/api_docs/python/tf/data/Dataset). For more information about the Dataset class, please see the [tutorial related to that subject](https://www.tensorflow.org/guide/datasets). An additional argument is required for this case: **```steps_per_epoch```** take an integer that will set the number of training step to run before moving to the next iteration. Finally, as ```Dataset``` is a TensorFlow-type, there is no need to specify the batch as it is wrapped in the ```Dataset``` instance:

In [7]:
dataset = tf.data.Dataset.from_tensor_slices((features, labels))
dataset = dataset.batch(32).repeat()

val_dataset = tf.data.Dataset.from_tensor_slices((val_features, val_labels))
val_dataset = val_dataset.batch(32).repeat()

import logging

try:
    model.fit(
        x=dataset,
        epochs=10,
        steps_per_epoch=30,
        validation_data=val_dataset,
        validation_steps=3
    )
except (ValueError, AttributeError) as e:
    logging.exception("Error while training model")

ERROR:root:Error while training model
Traceback (most recent call last):
  File "<ipython-input-7-d90b06a579a9>", line 15, in <module>
    validation_steps=3
  File "/home/valentin/anaconda3/envs/tensorflow/lib/python3.5/site-packages/keras/engine/training.py", line 955, in fit
    batch_size=batch_size)
  File "/home/valentin/anaconda3/envs/tensorflow/lib/python3.5/site-packages/keras/engine/training.py", line 754, in _standardize_user_data
    exception_prefix='input')
  File "/home/valentin/anaconda3/envs/tensorflow/lib/python3.5/site-packages/keras/engine/training_utils.py", line 90, in standardize_input_data
    data = [standardize_single_array(x) for x in data]
  File "/home/valentin/anaconda3/envs/tensorflow/lib/python3.5/site-packages/keras/engine/training_utils.py", line 90, in <listcomp>
    data = [standardize_single_array(x) for x in data]
  File "/home/valentin/anaconda3/envs/tensorflow/lib/python3.5/site-packages/keras/engine/training_utils.py", line 25, in standardize_si

#### Evaluation & Prediction ####

This is it! The model is finally ready to evaluate and predict some stuff! For that, the methods [```model.evaluate()```](https://www.tensorflow.org/api_docs/python/tf/keras/Model#evaluate) and [```model.predict()```](https://www.tensorflow.org/api_docs/python/tf/keras/Model#predict) will be used.

The evaluation test the model by taking *features* and *labels* as argument, and print the accuracy of the artificial intelligence:

In [8]:
# Use same dataset as for training
print(model.evaluate(features, labels, batch_size=32))

try:
    model.evaluate(dataset, steps=30)
except AttributeError:
    pass

# Use new dataset
eval_features = np.random.random((1000, 32))
eval_labels   = np.random.random((1000, 10))

print(model.evaluate(eval_features, eval_labels, batch_size=32))

1000/1000 [==============================] - 0s 29us/step
[11.499515228271484, 0.173]
1000/1000 [==============================] - 0s 34us/step
[11.530293426513673, 0.095]


Again, the loss and accuracy values show that the model is not well-trained (because of the randomly-generated datasets)

Now, let's see the model making predictions:

In [9]:
print(model.predict(features, batch_size=32))
print(model.predict(np.random.random((1, 32)), batch_size=32))

try:
    model.predict(dataset, steps=30)
except AttributeError:
    pass

[[0.11069123 0.10190696 0.09946288 ... 0.08772218 0.09612855 0.10351599]
 [0.09694731 0.12679659 0.09947062 ... 0.08486484 0.09469884 0.09670925]
 [0.08241725 0.1080385  0.09340771 ... 0.11289827 0.098519   0.0972159 ]
 ...
 [0.10156006 0.10767496 0.1047217  ... 0.09856814 0.09940234 0.10035412]
 [0.10589682 0.10590004 0.09201646 ... 0.09842134 0.0940019  0.09853192]
 [0.09517039 0.11262779 0.11501919 ... 0.07984426 0.09563505 0.11859612]]
[[0.10792921 0.11201751 0.10585713 0.09933899 0.09584751 0.09979638
  0.10214879 0.08704121 0.09579197 0.09423132]]


### Advanced Model ###

The [```tf.keras.Sequential```](https://www.tensorflow.org/api_docs/python/tf/keras/Sequential) class to generate a model is cute, but not very effective all the time. Depending on the problem, other type of model might be selected (for more information about Keras model, see the [functional API guide](https://keras.io/getting-started/functional-api-guide/))

[...]

### Save & Load ###

#### Weights ####

From now on, we created a model, train it, evaluate it, to finally make prediction. But imagine now that the training is an incredibly long process (this is the case in the real world), same for the evaluation phase. Do you believe that each time your computer or your server will train it from scratch at every startup? Of course, it would be a waste of time. Fortunatly for us, TensorFlow developers implemented a way to save and restore our model. In this example, my model will be saved in this folder:

In [10]:
model_path = "../../../../../res/saved_models/keras_tuto_model"

The TensorFlow team created the method [```model.save_weights()```](https://www.tensorflow.org/api_docs/python/tf/keras/Model#save_weights) to save the model weights, so you only have to run the training process once. Then, you can call [```model.load_weights()```](https://www.tensorflow.org/api_docs/python/tf/keras/Model#load_weights) to load your model every time you need to! Ideally, a small evaluation would be perfect to assess the loaded model. The code below described this very simple in only 3 lines of code:

In [11]:
import h5py

model.save_weights(model_path + "_weights")
model.load_weights(model_path + "_weights")

You may have see that the python package h5py is required. If you are on Debian/Ubuntu, you must run the following command:
```bash
sudo apt-get install libhdf5-dev
```
Then, the python package h5py must be installed in your environment. If you use pip, you can run this command:
```bash
pip install h5py
```
If you use an Anaconda environment, run that command:
```bash
conda install -n myenv h5py
```
Please the the [Stack Overflow forum about this subject for more details](https://stackoverflow.com/questions/43385565/importerrorsave-weights-requires-h5py).

The file created is a [TensorFlow checkpoint file](https://www.tensorflow.org/guide/checkpoints), but we can specify to save our model according to the Keras HDF5 file format:

In [12]:
try:
    model.save_weights(model_path + "_weights.h5", save_format="h5")
    model.load_weights(model_path + "_weights.h5")
except TypeError:
    logging.exception("Error while saving|loading model")

ERROR:root:Error while saving|loading model
Traceback (most recent call last):
  File "<ipython-input-12-92659b39e074>", line 2, in <module>
    model.save_weights(model_path + "_weights.h5", save_format="h5")
TypeError: save_weights() got an unexpected keyword argument 'save_format'


#### Configuration ####

Previously, we saved only the **weights** of the model, but now we will save the configuration only as a JSON or YAML format:

In [13]:
# Get the JSON string format of the model:
json_format = model.to_json()
print(json_format)

# Write it
with open(model_path + "_config.json", 'w') as f:
    f.write(json_format + "\n")

del json_format

# Read it
json_format = None
with open(model_path + "_config.json", 'r') as f:
    json_format = f.read()

assert json_format != None

new_model = keras.models.model_from_json(json_format)

{"config": [{"config": {"activity_regularizer": null, "units": 64, "kernel_regularizer": null, "name": "dense_1", "use_bias": true, "trainable": true, "activation": "relu", "dtype": "float32", "bias_constraint": null, "kernel_initializer": {"config": {"scale": 1.0, "mode": "fan_avg", "seed": null, "distribution": "uniform"}, "class_name": "VarianceScaling"}, "bias_regularizer": null, "bias_initializer": {"config": {}, "class_name": "Zeros"}, "kernel_constraint": null, "batch_input_shape": [null, 32]}, "class_name": "Dense"}, {"config": {"activity_regularizer": null, "units": 64, "kernel_regularizer": null, "name": "dense_2", "use_bias": true, "trainable": true, "activation": "relu", "bias_constraint": null, "kernel_initializer": {"config": {"scale": 1.0, "mode": "fan_avg", "seed": null, "distribution": "uniform"}, "class_name": "VarianceScaling"}, "bias_regularizer": null, "bias_initializer": {"config": {}, "class_name": "Zeros"}, "kernel_constraint": null}, "class_name": "Dense"}, {"c

Now, let's serialize the model in YAML:

In [14]:
# Get the YAML string format of the model:
yaml_format = model.to_yaml()
print(yaml_format)

# Write it
with open(model_path + "_config.yaml", 'w') as f:
    f.write(yaml_format + "\n")

del yaml_format

# Read it
yaml_format = None
with open(model_path + "_config.yaml", 'r') as f:
    yaml_format = f.read()

assert yaml_format != None

new_model = keras.models.model_from_yaml(yaml_format)

backend: tensorflow
class_name: Sequential
config:
- class_name: Dense
  config:
    activation: relu
    activity_regularizer: null
    batch_input_shape: !!python/tuple [null, 32]
    bias_constraint: null
    bias_initializer:
      class_name: Zeros
      config: {}
    bias_regularizer: null
    dtype: float32
    kernel_constraint: null
    kernel_initializer:
      class_name: VarianceScaling
      config: {distribution: uniform, mode: fan_avg, scale: 1.0, seed: null}
    kernel_regularizer: null
    name: dense_1
    trainable: true
    units: 64
    use_bias: true
- class_name: Dense
  config:
    activation: relu
    activity_regularizer: null
    bias_constraint: null
    bias_initializer:
      class_name: Zeros
      config: {}
    bias_regularizer: null
    kernel_constraint: null
    kernel_initializer:
      class_name: VarianceScaling
      config: {distribution: uniform, mode: fan_avg, scale: 1.0, seed: null}
    kernel_regularizer: null
    name: dense_2
    trainabl

#### Entire model ####

In most of the case, you will save the **entire model**, including the **weights**, the **configuration** and the **optimizer**:

In [15]:
model = keras.Sequential()
model.add(keras.layers.Dense(units=64, activation="relu", input_shape=(32,)))
model.add(keras.layers.Dense(units=64, activation="relu"))
model.add(keras.layers.Dense(units=10, activation="softmax"))
model.compile(
    optimizer=tf.train.AdamOptimizer(0.001),
    loss="categorical_crossentropy",
    metrics=["accuracy"]
)
model.fit(features, labels, batch_size=32, epochs=10)

model.save(model_path + ".h5")
model = keras.models.load_model(model_path + ".h5")

Epoch 1/10
1000/1000 [==============================] - 0s 363us/step - loss: 11.5878 - acc: 0.1120
Epoch 2/10
1000/1000 [==============================] - 0s 75us/step - loss: 11.5491 - acc: 0.1100
Epoch 3/10
1000/1000 [==============================] - 0s 81us/step - loss: 11.5381 - acc: 0.1430
Epoch 4/10
1000/1000 [==============================] - 0s 85us/step - loss: 11.5329 - acc: 0.1430
Epoch 5/10
1000/1000 [==============================] - 0s 80us/step - loss: 11.5294 - acc: 0.1620
Epoch 6/10
1000/1000 [==============================] - 0s 83us/step - loss: 11.5236 - acc: 0.1760
Epoch 7/10
1000/1000 [==============================] - 0s 84us/step - loss: 11.5185 - acc: 0.1590
Epoch 8/10
1000/1000 [==============================] - 0s 75us/step - loss: 11.5173 - acc: 0.1660
Epoch 9/10
1000/1000 [==============================] - 0s 74us/step - loss: 11.5114 - acc: 0.1840
Epoch 10/10
1000/1000 [==============================] - 0s 69us/step - loss: 11.5083 - acc: 0.1650


/home/valentin/anaconda3/envs/tensorflow/lib/python3.5/site-packages/keras/engine/saving.py:127: UserWarning: TensorFlow optimizers do not make it possible to access optimizer attributes or optimizer state after instantiation. As a result, we cannot save the optimizer as part of the model save file.You will have to compile your model again after loading it. Prefer using a Keras optimizer instead (see keras.io/optimizers).
  'TensorFlow optimizers do not '
/home/valentin/anaconda3/envs/tensorflow/lib/python3.5/site-packages/keras/engine/saving.py:270: UserWarning: No training configuration found in save file: the model was *not* compiled. Compile it manually.
  warnings.warn('No training configuration found in save file: '
